llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -5567,7 +5567,7 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
 
 bool SelectionDAG::isKnownNonNegativeFP(SDValue Op) const {
   if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(Op, true))
-    return !C1->isNegative();
+    return !C1->isNegative() && !C1->isNaN();
```

Comment:
I'm not sure this is appropriate since the result of the FABS case below could still be NaN. What is the exact spec of this function?